connector/connector_ldap.go

```diff
@@ -0,0 +1,387 @@
+)
+
+func init() {
+	RegisterConnectorConfigType(LDAPConnectorType, func() ConnectorConfig { return &LDAPConnectorConfig{} })
+}
+
+type LDAPConnectorConfig struct {
+	ID                   string        `json:"id"`
+	ServerHost           string        `json:"serverHost"`
+	ServerPort           uint16        `json:"serverPort"`
+	Timeout              time.Duration `json:"timeout"`
+	UseTLS               bool          `json:"useTLS"`
+	UseSSL               bool          `json:"useSSL"`
+	CertFile             string        `json:"certFile"`
+	KeyFile              string        `json:"keyFile"`
+	CaFile               string        `json:"caFile"`
+	SkipCertVerification bool          `json:"skipCertVerification"`
+	BaseDN               string        `json:"baseDN"`
+	NameAttribute        string        `json:"nameAttribute"`
+	EmailAttribute       string        `json:"emailAttribute"`
+	SearchBeforeAuth     bool          `json:"searchBeforeAuth"`
+	SearchFilter         string        `json:"searchFilter"`
+	SearchScope          string        `json:"searchScope"`
+	SearchBindDN         string        `json:"searchBindDN"`
+	SearchBindPw         string        `json:"searchBindPw"`
+	BindTemplate         string        `json:"bindTemplate"`
+	TrustedEmailProvider bool          `json:"trustedEmailProvider"`
+}
+
+func (cfg *LDAPConnectorConfig) ConnectorID() string {
+	return cfg.ID
+}
+
+func (cfg *LDAPConnectorConfig) ConnectorType() string {
+	return LDAPConnectorType
+}
+
+type LDAPConnector struct {
+	id                   string
+	idp                  *LDAPIdentityProvider
+	namespace            url.URL
+	trustedEmailProvider bool
+	loginFunc            oidc.LoginFunc
+	loginTpl             *template.Template
+}
+
+func (cfg *LDAPConnectorConfig) Connector(ns url.URL, lf oidc.LoginFunc, tpls *template.Template) (Connector, error) {
+	ns.Path = path.Join(ns.Path, httpPathCallback)
+	tpl := tpls.Lookup(LDAPLoginPageTemplateName)
+	if tpl == nil {
+		return nil, fmt.Errorf("unable to find necessary HTML template")
+	}
+
+	if cfg.UseTLS && cfg.UseSSL {
+		return nil, fmt.Errorf("Invalid configuration. useTLS and useSSL are mutual exclusive.")
+	}
+
+	if len(cfg.CertFile) > 0 && len(cfg.KeyFile) == 0 {
+		return nil, fmt.Errorf("Invalid configuration. Both certFile and keyFile must be specified.")
+	}
+
+	var nameAttribute, emailAttribute, bindTemplate string
+	if len(cfg.NameAttribute) > 0 {
+		nameAttribute = cfg.NameAttribute
+	} else {
+		nameAttribute = "cn"
+	}
+
+	if len(cfg.EmailAttribute) > 0 {
+		emailAttribute = cfg.EmailAttribute
+	} else {
+		emailAttribute = "mail"
+	}
+
+	if len(cfg.BindTemplate) > 0 {
+		if cfg.SearchBeforeAuth {
+			log.Warningf("bindTemplate not used when searchBeforeAuth specified.")
+		}
+		bindTemplate = cfg.BindTemplate
+	} else {
+		bindTemplate = "uid=%u,%b"
+	}
+
+	var searchScope int
+	if len(cfg.SearchScope) > 0 {
+		switch {
+		case strings.EqualFold(cfg.SearchScope, "BASE"):
```

Comment:
I noticed in your configuration you're using lower case values, but your comparing to upper case values. Perhaps you should `strings.ToLower(cfg.SearchScope)` value, and compare it to lowercase string literals (`"base"` instead of `"BASE"`). I might be misunderstanding how EqualFold works though.